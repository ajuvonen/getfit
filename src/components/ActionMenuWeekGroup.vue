<script setup lang="ts">
import {storeToRefs} from 'pinia';
import type {Training} from '@/types';
import useWeekDays from '@/hooks/weekdays';
import {useScheduleStore} from '@/stores/schedule';

const props = defineProps<{
  action: 'move' | 'copy';
  training: Training;
}>();

const scheduleStore = useScheduleStore();
const {settings, weeks} = storeToRefs(scheduleStore);
const {moveTraining, copyTraining} = scheduleStore;

const {getDisplayWeekNumber, getShortDate, weekdays} = useWeekDays();

const menuAction = props.action === 'move' ? moveTraining : copyTraining;
const menuIcon = props.action === 'move' ? 'mdi-arrow-all' : 'mdi-content-copy';
</script>
<template>
  <v-list-group>
    <template v-slot:activator="{props}">
      <v-list-item
        v-bind="props"
        :title="$t(`trainingCard.${action}`)"
        :prepend-icon="menuIcon"
      />
    </template>
    <v-list-group v-for="(week, weekIndex) in weeks" :key="week.id">
      <template v-slot:activator="{props}">
        <v-list-item
          v-bind="props"
          :title="$t('weekCalendar.weekTitle', [getDisplayWeekNumber(weekIndex)])"
        />
      </template>
      <v-list-item
        v-for="(day, dayIndex) in weekdays"
        :disabled="week.id === training.weekId && dayIndex === training.dayIndex"
        :key="day"
        :title="day"
        @click="menuAction(training, week.id, dayIndex)"
      >
        <template v-if="settings.startDate" #append>
          <span class="ml-4">{{ getShortDate(dayIndex, weekIndex) }}</span>
        </template>
      </v-list-item>
    </v-list-group>
  </v-list-group>
</template>
<style lang="scss" scoped></style>
