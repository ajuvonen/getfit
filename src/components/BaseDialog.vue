<script setup lang="ts">
import useScreen from '@/hooks/screen';

withDefaults(
  defineProps<{
    show: boolean;
    title: string;
    useFullScreen?: boolean;
  }>(),
  {
    useFullScreen: true,
  },
);

const {isSmallScreen} = useScreen();
</script>
<template>
  <v-dialog
    :model-value="show"
    :width="!isSmallScreen ? 'auto' : undefined"
    :fullscreen="useFullScreen && isSmallScreen"
    :transition="isSmallScreen ? 'dialog-bottom-transition' : false"
    :min-width="!isSmallScreen ? 500 : undefined"
    persistent
  >
    <v-card>
      <v-card-title>{{ title }}</v-card-title>
      <v-card-text>
        <slot name="content" />
      </v-card-text>
      <v-card-actions><slot name="actions" /></v-card-actions>
    </v-card>
  </v-dialog>
</template>
<style scoped></style>
