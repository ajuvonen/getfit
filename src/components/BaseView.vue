<script setup lang="ts">
import useScreenSize from '@/hooks/screenSize';

defineProps<{
  title: string;
  guide: string;
}>();

const {isSmallScreen, isMediumScreen} = useScreenSize();
</script>
<template>
  <v-card
    color="rgba(255,255,255,0.9)"
    class="mb-4"
    :rounded="isSmallScreen || isMediumScreen ? 0 : 'rounded'"
  >
    <v-card-title class="d-print-none">
      <h1 class="text-h4">{{ title }}</h1>
    </v-card-title>
    <v-card-text>
      <p class="d-flex align-center d-print-none mb-4">
        <v-icon icon="mdi-information-box-outline" />
        <span class="ml-4">{{ guide }}</span>
      </p>
      <slot name="content" />
    </v-card-text>
    <v-card-actions v-if="$slots.actions">
      <slot name="actions" />
    </v-card-actions>
  </v-card>
</template>