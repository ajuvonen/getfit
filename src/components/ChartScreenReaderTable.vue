<script lang="ts" setup>
defineProps<{
  id: string,
  title: string,
  labels: string[],
  data: number[],
}>();
</script>
<template>
  <table class="d-sr-only" :id="id">
    <caption>{{ title }}</caption>
    <thead>
      <tr>
        <th v-for="label in labels" :key="label">{{ label }}</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td v-for="count in data" :key="count">{{ count }}</td>
      </tr>
    </tbody>
  </table>
</template>
