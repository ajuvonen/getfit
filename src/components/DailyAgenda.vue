<script lang="ts" setup>
import {computed} from 'vue';
import {storeToRefs} from 'pinia';
import {DateTime} from 'luxon';
import {useScheduleStore} from '@/stores/schedule';
import TrainingCard from '@/components/TrainingCard.vue';

const scheduleStore = useScheduleStore();
const {weeks, settings, getTrainingsCount, getCompletedTrainingsCount} = storeToRefs(scheduleStore);

const dailyTrainings = computed(() => {
  const startDate = DateTime.fromJSDate(settings.value.startDate!);
  const difference = Math.ceil(startDate.diffNow('weeks').weeks);
  if (difference < 1) {
    const targetWeek = weeks.value[Math.abs(difference)];
    if (targetWeek) {
      const targetDay = (DateTime.now().weekday - (settings.value.startsOnSunday ? 0 : 1)) % 7;
      return targetWeek.trainings.filter(({dayIndex}) => dayIndex === targetDay);
    }
  }
  return [];
});
</script>
<template>
  <v-progress-circular
    :model-value="(getCompletedTrainingsCount / getTrainingsCount) * 100"
    class="mx-auto"
    size="150"
    width="15"
  >
    <v-fab-transition>
      <div v-if="getTrainingsCount > getCompletedTrainingsCount" class="d-flex flex-column">
        <span class="text-h4 daily-agenda__progress">{{ getCompletedTrainingsCount }} / {{ getTrainingsCount }}</span>
        <span class="text-caption text-uppercase">{{ $t('home.progress') }}</span>
      </div>
      <v-icon v-else class="daily-agenda__completed-icon" icon="$trophyOutline" size="70" />
    </v-fab-transition>
  </v-progress-circular>
  <h1 class="mt-10 text-h3 text-center">
    {{ $t('home.dailyAgenda') }}
  </h1>
  <div class="daily-agenda has-scroll mt-10 d-flex">
    <TrainingCard
      v-for="training in dailyTrainings"
      :key="training.id"
      :training="training"
      simple
    />
    <p v-if="!dailyTrainings.length" class="mx-auto text-h6">{{ $t('home.noTrainings') }}</p>
  </div>
</template>
<style lang="scss" scoped>
.daily-agenda {
  gap: 1rem;
  overflow-x: scroll;
}

.training-card__container:first-child {
  margin-left: auto;
}

.training-card__container:last-child {
  margin-right: auto;
}
</style>
