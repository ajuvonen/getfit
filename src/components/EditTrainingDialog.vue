<script setup lang="ts">
import {computed} from 'vue';
import {storeToRefs} from 'pinia';
import {useI18n} from 'vue-i18n';
import {useVuelidate} from '@vuelidate/core';
import {required, between, maxLength, integer, helpers} from '@vuelidate/validators';
import type {VForm} from 'vuetify/components';
import {Intensity} from '@/types';
import {useScheduleStore} from '@/stores/schedule';
import {useAppStateStore} from '@/stores/appState';
import {ACTIVITIES} from '@/constants';
import {getIntensityColor, getIcon, getValidationErrors, decimalRegex} from '@/utils';
import useScreenSize from '@/hooks/screenSize';
import useLocalizedActivities from '@/hooks/localizedActivities';

const scheduleStore = useScheduleStore();
const {settings} = storeToRefs(scheduleStore);
const {addOrEditTraining} = scheduleStore;

const appStateStore = useAppStateStore();
const {trainingDialogOpen, trainingData} = storeToRefs(appStateStore);

const {t} = useI18n();

const {isSmallScreen} = useScreenSize();

const tickLabels = computed(() => ({
  [Intensity.LIGHT]: t('intensities.LIGHT'),
  [Intensity.NORMAL]: t('intensities.NORMAL'),
  [Intensity.MEDIUM]: t('intensities.MEDIUM'),
  [Intensity.HEAVY]: t('intensities.HEAVY'),
}));

const {localizedAvailableActivities} = useLocalizedActivities();

const handleSave = async () => {
  if (await v$.value.$validate()) {
    addOrEditTraining(trainingData.value);
    resetAndClose();
  }
};

const rules = computed(() => ({
  title: {maxLength: maxLength(30)},
  location: {maxLength: maxLength(40)},
  description: {maxLength: maxLength(2000)},
  duration: {
    required,
    between: between(0, settings.value.unitOfTime === 'm' ? 360 : 6),
    precision:
      settings.value.unitOfTime === 'm'
        ? integer
        : helpers.withMessage(t('errors.invalidPrecision'), decimalRegex),
  },
  activity: {required},
}));

const v$ = useVuelidate(rules, trainingData);

const resetAndClose = () => {
  v$.value.$reset();
  trainingDialogOpen.value = false;
};
</script>
<template>
  <v-dialog
    v-model="trainingDialogOpen"
    :width="!isSmallScreen ? 'auto' : undefined"
    :fullscreen="isSmallScreen"
    :transition="isSmallScreen ? 'dialog-bottom-transition' : false"
    :min-width="!isSmallScreen ? 500: undefined"
    persistent
  >
    <v-card>
      <v-card-title>{{ $t('editTraining.title') }}</v-card-title>
      <v-card-text>
        <v-form>
          <v-label for="edit-training-activity">{{ $t('editTraining.activity') }}</v-label>
          <v-select
            id="edit-training-activity"
            v-model="trainingData.activity"
            :error-messages="getValidationErrors(v$.activity)"
            :items="localizedAvailableActivities"
            eager
            data-test-id="edit-training-activity"
            @input="v$.activity.$touch"
            @blur="v$.activity.$touch"
          >
            <template #selection="{item: {title, value}}">
              <div v-if="value">
                <v-icon :icon="getIcon(ACTIVITIES, value)" />
                {{ title }}
              </div>
            </template>
            <template #item="{item: {title, value}, props}">
              <v-list-item v-bind="props" :data-test-id="`edit-training-activity-${value}`">
                <template #title>
                  <v-icon :icon="getIcon(ACTIVITIES, value)" />
                  {{ title }}
                </template>
              </v-list-item>
            </template>
          </v-select>
          <v-text-field
            v-model="trainingData.title"
            :label="t('editTraining.trainingTitle')"
            :error-messages="getValidationErrors(v$.title)"
            maxlength="30"
            counter
            data-test-id="edit-training-title"
            @input="v$.title.$touch"
            @blur="v$.title.$touch"
          />
          <v-text-field
            v-model="trainingData.location"
            :label="t('editTraining.location')"
            :error-messages="getValidationErrors(v$.location)"
            maxlength="40"
            counter
            data-test-id="edit-training-location"
            @input="v$.location.$touch"
            @blur="v$.location.$touch"
          />
          <v-text-field
            v-model.number="trainingData.duration"
            :error-messages="getValidationErrors(v$.duration)"
            :suffix="settings.unitOfTime"
            :label="t('editTraining.duration')"
            class="edit-training-duration"
            type="number"
            data-test-id="edit-training-duration"
            @input="v$.duration.$touch"
            @blur="v$.duration.$touch"
          />
          <v-textarea
            v-model="trainingData.description"
            :label="t('editTraining.description')"
            :error-messages="getValidationErrors(v$.description)"
            clearable
            maxlength="2000"
            no-resize
            counter
            data-test-id="edit-training-description"
            @input="v$.description.$touch"
            @blur="v$.description.$touch"
          />
          <v-label for="edit-training-intensity">{{ $t('editTraining.intensity') }}</v-label>
          <v-slider
            id="edit-training-intensity"
            v-model="trainingData.intensity"
            :ticks="tickLabels"
            :color="getIntensityColor(trainingData.intensity)"
            max="3"
            step="1"
            show-ticks="always"
            tick-size="4"
            data-test-id="edit-training-intensity"
          ></v-slider>
        </v-form>
      </v-card-text>
      <v-card-actions>
        <v-btn
          prepend-icon="mdi-content-save"
          variant="outlined"
          color="primary"
          data-test-id="edit-training-save-button"
          @click="handleSave()"
          >{{ $t('editTraining.save') }}</v-btn
        >
        <v-btn
          prepend-icon="mdi-close"
          data-test-id="edit-training-close-button"
          variant="tonal"
          @click="resetAndClose"
          >{{ $t('editTraining.closeDialog') }}</v-btn
        >
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<style lang="scss" scoped>
:deep(.edit-training-duration .v-input__control) {
  max-width: 200px;
  input[type='number'] {
    appearance: textField;
  }
}
</style>
