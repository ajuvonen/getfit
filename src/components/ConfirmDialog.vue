<script setup lang="ts">
import {storeToRefs} from 'pinia';
import {useAppStateStore} from '@/stores/appState';
import BaseDialog from '@/components/BaseDialog.vue';

const {confirmDialogOpen, confirmText, confirmAction} = storeToRefs(useAppStateStore());
</script>
<template>
  <BaseDialog
    :open="confirmDialogOpen"
    :use-full-screen="false"
    :title="$t('general.confirmTitle')"
  >
    <template #content>{{ confirmText }}</template>
    <template #actions>
      <v-btn
        color="primary"
        variant="outlined"
        prepend-icon="mdi-check"
        data-test-id="confirm-dialog-confirm-button"
        @click="confirmAction(), (confirmDialogOpen = false)"
        >{{ $t('general.confirm') }}</v-btn
      >
      <v-btn
        prepend-icon="mdi-close"
        data-test-id="confirm-dialog-cancel-button"
        @click="confirmDialogOpen = false"
        >{{ $t('general.cancel') }}</v-btn
      >
    </template>
  </BaseDialog>
</template>
