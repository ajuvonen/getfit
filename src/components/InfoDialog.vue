<script setup lang="ts">
import BaseDialog from '@/components/BaseDialog.vue';

defineProps<{
  show: boolean;
  title: string;
}>();

defineEmits(['close']);
</script>
<template>
  <BaseDialog :show="show" :use-full-screen="false" :title="title">
    <template #content>
      <slot name="content" />
    </template>
    <template #actions>
      <v-btn
        color="primary"
        prepend-icon="mdi-close"
        data-test-id="info-dialog-close-button"
        @click="$emit('close')"
        >{{ $t('general.close') }}</v-btn
      >
    </template>
  </BaseDialog>
</template>
