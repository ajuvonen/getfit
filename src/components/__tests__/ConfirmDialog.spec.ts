import {describe, it, expect, beforeEach, vi} from 'vitest';
import {mount} from '@vue/test-utils';
import {VCard} from 'vuetify/components';
import {useAppStateStore} from '@/stores/appState';
import ConfirmDialog from '@/components/ConfirmDialog.vue';

describe('ConfirmDialog', () => {
  let appStateStore: ReturnType<typeof useAppStateStore>;

  beforeEach(() => {
    appStateStore = useAppStateStore();
  });

  it('mounts', () => {
    appStateStore.confirmText = 'Are you sure?';
    appStateStore.confirmDialogOpen = true;
    const wrapper = mount(ConfirmDialog);
    expect(wrapper.findComponent(VCard).html()).toMatchSnapshot();
  });

  it('confirm works', async () => {
    appStateStore.confirmText = 'Are you sure?';
    appStateStore.confirmDialogOpen = true;
    appStateStore.confirmAction = vi.fn();
    const wrapper = mount(ConfirmDialog);
    await wrapper.findComponent(VCard).findByTestId('confirm-dialog-confirm-button').trigger('click');
    expect(appStateStore.confirmAction).toHaveBeenCalledOnce();
    expect(appStateStore.confirmDialogOpen).toBe(false);
  });

  it('cancel works', async () => {
    appStateStore.confirmText = 'Are you sure?';
    appStateStore.confirmDialogOpen = true;
    appStateStore.confirmAction = vi.fn();
    const wrapper = mount(ConfirmDialog);
    await wrapper.findComponent(VCard).findByTestId('confirm-dialog-cancel-button').trigger('click');
    expect(appStateStore.confirmAction).toHaveBeenCalledTimes(0);
    expect(appStateStore.confirmDialogOpen).toBe(false);
  });
});
