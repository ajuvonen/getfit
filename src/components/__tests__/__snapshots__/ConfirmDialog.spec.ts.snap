// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmDialog > mounts 1`] = `
"<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-title">Please confirm</div>
  <div class="v-card-text">Are you sure?</div>
  <div class="v-card-actions"><button type="button" class="v-btn v-btn--slim v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-outlined" data-test-id="confirm-dialog-confirm-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-check mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Confirm</span>
      <!---->
      <!---->
    </button><button type="button" class="v-btn v-btn--slim v-theme--light v-btn--density-default v-btn--size-default v-btn--variant-text" data-test-id="confirm-dialog-cancel-button"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span><span class="v-btn__prepend"><i class="mdi-close mdi v-icon notranslate v-theme--light v-icon--size-default" aria-hidden="true"></i></span><span class="v-btn__content" data-no-activator="">Cancel</span>
      <!---->
      <!---->
    </button></div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
